torch_xla/csrc/tensor.cpp

```diff
@@ -1114,9 +1122,15 @@ XLATensor::SyncTensorCollection XLATensor::CollectSyncTensors(
   TF_VLOG(4) << "Waiting on device barrier for device " << coll.device
              << " done!";
   for (size_t i = 0; i < tensors.size(); ++i) {
+    xla::int64 tensor_id = tensors[i].GetUniqueId();
+    if (tensor_ids.count(tensor_id) > 0) {
+      continue;
+    }
+    tensor_ids.insert(tensor_id);
     if (tensors[i].CurrentXlaData() == nullptr) {
       ir::Value ir_value = tensors[i].CurrentIrValue();
-      if (ir_value) {
```

Comment:
The lines here and below should not be changed.
If a tensor has no XLA data, and no IR value, it must have pytorch Tensor data.

Reply:
Right, but it is possible that a tensor has no `xla_data` but has `ir_value` and `tensor_data`. Like the example I posted 

```
tensorID: 3
ir value: s64[3]{0} xla::device_data, location=<module>@example.py:20, device=CPU:0
with xla data: False
tensor data: 
 2
 4
 6
[ CPULongType{3} ]
```

My understanding is that in this case we don't want to sync, hence I added `!tensor_data` to the if check

Reply:
The current bug is that, we will go ahead and sync this tensor since it has no `XLA data` and has `ir_value` so we will go ahead and try to sync this tensor, how ever in the `getTensor` function we have

https://github.com/pytorch/xla/blob/5cb463298943df472936b45d727b48501e347dd5/torch_xla/csrc/tensor.cpp#L985

and we will use `tensor_value` directly instead of from the `sync_list` and create a index mismatched for other tensors.

Reply:
Thanks @JackCaoG ! let me try to repro it later today or tomorrow. IR value and Tensor data should not be possible.

Reply:
@davidel Sounds good! The tensor with `tensor_data` and `ir_value` is `y` in my example. It is the result of a default OP(I removed the lowering of `add` like in [here](https://github.com/pytorch/xla/commit/a3387469b71718c1112acafe08ce48d139763436))

Reply:
@JackCaoG Give https://gist.github.com/davidel/f499660efa1fb161f7d6cb066eeda925 a try.
Essentially we need to have the indices created by the collect API to be the source of truth down the code paths.

Stopping my VM now ... I must have used almost $50 today! 😄